src/util.h

```diff
@@ -100,6 +88,41 @@ typedef struct IntRect {
 #undef min
 #undef max
 
+typedef struct FloatRect {
+	float x;
+	float y;
+	float w;
+	float h;
+} FloatRect;
+
+typedef struct IntRect {
+	int x;
+	int y;
+	int w;
+	int h;
+} IntRect;
+
+typedef struct Ellipse {
+	complex origin;
+	complex axes; // NOT half-axes!
+	double angle;
+} Ellipse;
+
+typedef struct LineSegment {
+	complex a;
+	complex b;
+} LineSegment;
+
+typedef struct Circle {
+	complex origin;
+	double radius;
+} Circle;
+
+typedef struct Rect {
+	complex top_left;
+	complex bottom_right;
+} Rect;
+
```

Comment:
I bet ellipse and circle will never be used outside of collision code.

Reply:
Probably, but we already have collision code in more than one place (projectiles and lasers), and we may want something more sophisticated for bosses/enemies later - Yuyuko's huge-ass shootable banner in PCB comes to mind.